### ISSUE TITLE
[Bug]: cron.php executes successfully but Nextcloud says cronjob hasn't run

### ISSUE THREAD

**Issue by hegedussz** [nextcloud/server#30899]
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

I saw this message: 
Background jobs  Some jobs didn’t execute since yesterday. 
Please consider switching to system cron.

### Steps to reproduce

1. working instance 23.0
2. set the webcron  
3 cron being run by webcron 

### Expected behavior

The cron has seemingly run successfully, so I should not see this message in the UI? Can't see any issues in the Nextcloud log either.

### Installation method

Official Docker image

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 8.0

### Web server

Apache (supported)

### Database engine version

MariaDB

### Is this bug present after an update or on a fresh install?

Fresh Nextcloud Server install

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### What user-backends are you using?

- [ ] Default user-backend _(database)_
- [X] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```shell
{
    "system": {
        "htaccess.RewriteBase": "\/",
        "memcache.local": "\\OC\\Memcache\\APCu",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 6379
        },
        "apps_paths": [
            {
                "path": "\/var\/www\/html\/apps",
                "url": "\/apps",
                "writable": false
            },
            {
                "path": "\/var\/www\/html\/custom_apps",
                "url": "\/custom_apps",
                "writable": true
            }
        ],
        "overwriteprotocol": "https",
        "objectstore": {
            "class": "\\OC\\Files\\ObjectStore\\S3",
            "arguments": {
                "bucket": "nextcloud-prod-01",
                "key": "***REMOVED SENSITIVE VALUE***",
                "secret": "***REMOVED SENSITIVE VALUE***",
                "region": "hu-central-1",
                "hostname": "staging.s3.einfra.hu",
                "port": "",
                "objectPrefix": "urn:oid:",
                "autocreate": false,
                "use_ssl": true,
                "use_path_style": true,
                "legacy_auth": false
            }
        },
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "localhost",
            "nextcloud-prod-01.prod.caas.einfra.hu",
            "nextcloud.dinpi.hu"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "23.0.0.10",
        "overwrite.cli.url": "https:\/\/localhost",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "maintenance": false,
        "mail_smtpmode": "smtp",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_sendmailmode": "smtp",
        "mail_smtpport": "25",
        "loglevel": 0,
        "debug": true,
        "ldapProviderFactory": "OCA\\User_LDAP\\LDAPProviderFactory",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***"
    }
}
```

### List of activated Apps

```shell
Enabled:
  - accessibility: 1.9.0
  - activity: 2.15.0
  - admin_audit: 1.13.0
  - calendar: 3.0.5
  - cloud_federation_api: 1.6.0
  - comments: 1.13.0
  - contacts: 4.0.7
  - contactsinteraction: 1.4.0
  - dashboard: 7.3.0
  - dav: 1.21.0
  - event_update_notification: 1.4.0
  - federatedfilesharing: 1.13.0
  - federation: 1.13.0
  - files: 1.18.0
  - files_pdfviewer: 2.4.0
  - files_rightclick: 1.2.0
  - files_sharing: 1.15.0
  - files_trashbin: 1.13.0
  - files_versions: 1.16.0
  - files_videoplayer: 1.12.0
  - firstrunwizard: 2.12.0
  - group_default_quota: 0.1.3
  - logreader: 2.8.0
  - lookup_server_connector: 1.11.0
  - mail: 1.11.6
  - nextcloud_announcements: 1.12.0
  - notifications: 2.11.1
  - oauth2: 1.11.0
  - password_policy: 1.13.0
  - photos: 1.5.0
  - privacy: 1.7.0
  - provisioning_api: 1.13.0
  - recommendations: 1.2.0
  - serverinfo: 1.13.0
  - settings: 1.5.0
  - sharebymail: 1.13.0
  - support: 1.6.0
  - survey_client: 1.11.0
  - systemtags: 1.13.0
  - text: 3.4.0
  - theming: 1.14.0
  - twofactor_backupcodes: 1.12.0
  - updatenotification: 1.13.0
  - user_ldap: 1.13.0
  - user_status: 1.3.1
  - viewer: 1.7.0
  - weather_status: 1.3.0
  - workflowengine: 2.5.0
Disabled:
  - circles
  - encryption
  - files_external
```

### Nextcloud Signing status

```shell
No errors have been found.
```

### Nextcloud Logs

```shell
{"reqId":"j1YyIpzI8Jd67gESXKwa","level":0,"time":"2022-01-28T18:55:09+00:00","remoteAddr":"255.255.255.255","user":"--","app":"cron","method":"GET","url":"/cron.php","message":"Finished OCA\\Mail\\BackgroundJob\\TrainImportanceClassifierJob job with ID 2247 in 0 seconds","userAgent":"curl/7.81.0-DEV","version":"23.0.0.10","id":"61f43c0dcdef9"}

{"reqId":"j1YyIpzI8Jd67gESXKwa","level":0,"time":"2022-01-28T18:55:09+00:00","remoteAddr":"255.255.255.255","user":"--","app":"cron","method":"GET","url":"/cron.php","message":"Run OCA\\Mail\\BackgroundJob\\TrainImportanceClassifierJob job with ID 2247","userAgent":"curl/7.81.0-DEV","version":"23.0.0.10","id":"61f43c0dcdf95"}

{"reqId":"bXxjA2YbyH09Kix5PzKQ","level":0,"time":"2022-01-28T18:50:12+00:00","remoteAddr":"255.255.255.255","user":"--","app":"cron","method":"GET","url":"/cron.php","message":"Run OCA\\Mail\\BackgroundJob\\SyncJob job with ID 2502","userAgent":"curl/7.81.0-DEV","version":"23.0.0.10","id":"61f43ae53d1c1"}
```

### Additional info

Webcron container: 
Image: curlimages/curl:7.81.0
Cron Schedule: */5 * * * *

Command: curl -L -s https://nextcloud.ltd/cron.php
Result: {"status":"success"}

When I run cron.php manually inside the container it is successful (sudo -u www-data PHP_MEMORY_LIMIT=512M php /var/html/cron.php)

**Comment by darktweek** [comment]
I have same problem

**Comment by szaimen** [comment]
Yes, webcron is known to be unreliable. Please switch to system cron instead! https://docs.nextcloud.com/server/stable/admin_manual/configuration_server/background_jobs_configuration.html?highlight=cron#cron

**Comment by hegedussz** [comment]
> Yes, webcron is known to be unreliable. Please switch to system cron instead! https://docs.nextcloud.com/server/stable/admin_manual/configuration_server/background_jobs_configuration.html?highlight=cron#cron

Unfortunately the offical Nextcloud docker image does not use the system cron.

**Comment by szaimen** [comment]
you can either add an addditional container for this: https://github.com/nextcloud/docker/blob/dcf058fb559e7918cd77f5f1bb0dfdbcee813571/.examples/docker-compose/insecure/postgres/fpm/docker-compose.yml#L40-L48 or add a cronjob like this:

```
# Edit the root crontab
sudo crontab -e -u root

# Add this line to the file:
*/5 * * * * docker exec -it nextcloud-container-name php occ
# (of course you need to change 'nextcloud-container-name' to the correct name of your container)

# close and save the file -> done.
```

**Comment by koehn** [comment]
Any chance you'll update the documentation and remove the feature, or fix the bug?

**Comment by portno12** [comment]
the official helm chart also uses webcron and will need fixed along with the docs
https://github.com/nextcloud/helm/blob/master/charts/nextcloud/values.yaml#L281-L322

**Comment by koehn** [comment]
@szaimen This bug shouldn’t be closed; if a documented feature “is known to be unreliable” then either make it reliable or remove the feature. Leaving it documented and unreliable means it’s just a bug that you don’t want to fix.

**Comment by timkgh** [comment]
Related https://github.com/nextcloud/server/issues/31725

**Comment by charims** [comment]
I had this same issue, but it turned out I was really behind on cron jobs after the upgrade, so despite getting `{"status":"success"}` from webcron, it was taking a long time to catch up. From what I understand, since webcron only handles one job per-call, if you are really behind, it can take an astronomical amount of time to catch up. 

I ran the following sql query to check the status of the jobs:

```
select id, class, last_run, last_checked, execution_duration from oc_jobs order by last_run;
```
Most of mine had a last_run of 0.

So I wanted to see if webcron was doing anything, so i started to curl the cron.php endpoint. After a few minutes of calling it repeatedly it caught up. Now I just need to watch to see if it keeps up or if I need to change the frequency its called.

Running the below may help you get caught up on cron's rather quickly. Be careful, this may put some load on your system.

```
while true; do curl -s https://<domain>/cron.php; done
```
You can run this with a sleep statement to prevent causing too much system load:

```
while true; do curl -s https://<domain>/cron.php; sleep 5s; done

```
I knew I was done when the basic settings page had the green status for background jobs, and all the jobs had a recent last_run in the database.

**Comment by timkgh** [comment]
@charims I think it will fall behind again. I have a very small instance, system cron finishes quickly, then webcron is green for a while but after a few days it turns red again. Maybe webcron only runs a subset of all the jobs or it keeps failing on the same one but still returns `success`.

I switched to system cron for now, though I'd rather not, I don't like these cron jobs run by the `www-data` user.

**Comment by pneuschwander** [comment]
Could it be related to the value of `lastcron`?

After each successful cron run, the value of lastcron should be updated: https://github.com/nextcloud/server/blob/13dbad0080f31a5ae816779685070e8f94235d5a/cron.php#L180

That value determines whether the error message is shown in the UI:
https://github.com/nextcloud/server/blob/master/apps/settings/templates/settings/admin/server.php#L42 and the following lines.

The error message related to "Please consider switching to system cron" seems to not depend on a specific outstanding cron task. It seems to rather be based on the date of the last successful cron run.

The value of lastcron seems to be fetched for the UI at:
https://github.com/nextcloud/server/blob/b0fbcccfe66474d79586001ce9509d346919ae74/apps/settings/lib/Settings/Admin/Server.php#L75

So I'd assume that either their is something that prevents the update of lastcron value or that the retrieval of lastcron for the UI is not working as expected.

Could someone who is affected by this try to inspect the lastcron value (not sure where it is stored; maybe in a database table?) and check whether it is updated / fresh?

If it is updated/fresh: Maybe there is a problem to get the current value for the UI

If it is not updated / not fresh: Maybe there is a problem to persist the change or the line in cron.php is not reached (but then there should some exception be logged according to cron.php code)

Checking whether `lastcron` value is updated might help to better understand what is happening / where things go wrong.

**Comment by pneuschwander** [comment]
I think I found out what is going on. 
@fuse314 already referred to the different handling in cron.php based on whether execution is via webcron or cron/CLI.

![image](https://user-images.githubusercontent.com/8546924/164077224-7e0519ff-f9f7-42fd-b1b4-3ccbc0b37702.png)

The jobs with time_sensitive == 0 are the ones that won't be executed.

https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/background_jobs_configuration.html?highlight=cron#maintenance-window-start

has been introduced.

https://github.com/nextcloud/server/blob/396db12e8e97af033cff6dc0adaf13ee78ec5c3b/config/config.sample.php#L1238_L1250

The feature to respect `maintenance_window_start` is only part of the CLI option in cron.php.

https://github.com/nextcloud/server/blob/13dbad0080f31a5ae816779685070e8f94235d5a/cron.php#L114-L144

The getNext method of joblist has a new parameter with default = true.

https://github.com/nextcloud/server/blob/13dbad0080f31a5ae816779685070e8f94235d5a/lib/private/BackgroundJob/JobList.php#L206

https://github.com/nextcloud/server/blob/13dbad0080f31a5ae816779685070e8f94235d5a/lib/private/BackgroundJob/JobList.php#L215-L217

Webcron fetches job without specifying the parameter, fallback to true default parameter value and will never trigger time_sentitive == 0 jobs

https://github.com/nextcloud/server/blob/13dbad0080f31a5ae816779685070e8f94235d5a/cron.php#L170

One might want to consider changing the default value for the parameter or to adapt the "else" part of cron.php to consider the new parameter / feature.

**Comment by fuse314** [comment]
I modified cron.php to handle cron and webcron calls the same way:
My instance is now running [this version of cron.php.](https://github.com/fuse314/nextcloud_server/blob/webcron/cron.php)
So far - running only for a few hours - everything looks good, though each call to cron.php now takes a few seconds instead of finishing "immediately" before the change.
I suspect that when long running tasks will be handled, the webcron call might run into a timeout.

I will keep an eye on my small docker instance over the next days/weeks. If anyone here feels adventurous, you are welcome to try this out for yourself.

If I don't run into problems in the next week, I will open a pull request with this change.

**Comment by iordi154** [comment]
I wanted to try but ran Intro a user problem in my webhosted instance:
“ Console has to be executed with the user that owns the file config/config.php 
Current user id: 1  
id of config.php: 2“

**Comment by iordi154** [comment]
> Webcron fetches job without specifying the parameter, fallback to true default parameter value and will never trigger time_sentitive == 0 jobs
> 
> https://github.com/nextcloud/server/blob/13dbad0080f31a5ae816779685070e8f94235d5a/cron.php#L170
> 
> One might want to consider changing the default value for the parameter or to adapt the "else" part of cron.php to consider the new parameter / feature.

I can confirm: Manually changing time_sensitive parameters in DB = 1 immediately triggers all jobs and eliminate the red light in NC configuration menu => „everything OK“
However, the jobs fall back to time_sensitive = 0 so everything will start from zero again.

Sadly I don‘t know, what has to be done to change this behaviour.

**Comment by charims** [comment]
i've been running with modified cron.php from @fuse314 for a couple days, it got me caught back up after jobs ran behind again, and now it seems to be keeping up.

**Comment by pneuschwander** [comment]
I think https://github.com/nextcloud/server/pull/30945 introduced that changes related to `time_sensitive` jobs. Maybe @nickvergessen can add something to the discussion.

**Comment by nickvergessen** [comment]
See my comment in https://github.com/nextcloud/server/pull/32107#discussion_r856931358
We accidentally put the wrong default value in place for instances which use webcron or ajax cron jobs. Add `false` inside the `getNext()` and it should work again.

**Comment by k-popov** [comment]
I've created another #32109 with just default value changed to get **any** (time-sensitive or not) job from the list.

#32107 has been closed but still I don't understand why webcron should not execute time-sensitive jobs only during maintenance hours. I would be grateful for explanation why is that. I understand that this could be just maintainer's will but would be good to have rationale behind the decision.

What option is proposed for webcron to run time-consuming jobs during maintenance hours.

**Comment by nickvergessen** [comment]
Ill give a more detailed response to that tomorrow from desk. Too hard to put it together o. The phone in my free time 😅

**Comment by nickvergessen** [comment]
> What option is proposed for webcron to run time-consuming jobs during maintenance hours.

TL;DR: Only cron is suitable for non-single user instances

Background information:
* We try to set the php execution time to 1h, but in most environments this is prohibited and default execution times for web requests is limited, while unlimited for CLI:
https://github.com/nextcloud/server/blob/7d272c54d013538746d6731097ec37f360effb5d/lib/base.php#L631-L635
* To reduce the risk of breaking web-based crons (`ajax` and `webcron`) both only run 1 job per execution:
https://github.com/nextcloud/server/blob/3fd55cb4df80b6434cc5f028f80f1419b3f79202/cron.php#L162-L177
    * While CLI based `cron` loops and starts taking new jobs until it ran 14 minutes (by minute 15 3 other processes should run in parallel already)
      https://github.com/nextcloud/server/blob/3fd55cb4df80b6434cc5f028f80f1419b3f79202/cron.php#L144-L161
* With webcron running every 15 (very old recommendation) or 5 (recommended since some time) there are only 20 (in case of 15 min interval) or 60 (in case of 5 min interval) executions inside the maintenance window. So it's much too risky to only run jobs in that time.
* Ajax cron would even be killed completely mostlikely as the jobs would only run when no one is there, but when no one is there no one triggers the ajax requests to run a job.

Taking the numbers from the second last point a step further:
* 15min or 5min interval make 96 or 288 jobs that can be executed per day.
* I don't know if you ever checked your database, but e.g. on our company instance we have 722 jobs in the table right now
* Only 288 can be worked on today ~40%
* 434 jobs will not be executed today and have priority tomorrow

On my 2 user home installation I have 64 jobs, but there are no trashbin or version expirations queued, so they would manage to execute all together today.

**Summary:** Using ajax/webcron is only suitable for low-usage instances like 1-2 user home installs with passive sync behaviour. Anything else that requires serious cron execution to deliver notifications, emails, federation, .... must use `cron` to perform.
To perform even better on instances with many (> 1k) users we added this setting to allow admins to define load reduced hours so that some things which are not time critical can be moved to that time. We have to see how this develops as it reduces the load during the time people are online, but could e.g. heavily increase the load during the off-work times and might even be too much or something when the jobs are concentrated all into the same 4h window. But only reality from very big instances will show if this is a problem or better than the previous behaviour.

The breakage with the wrong default value was of course never intended. It's fixed now and will come with the next updates to 23 and 22 I assume.

**Comment by nickvergessen** [comment]
Also send a pull request to update the documentation on that a bit:
https://github.com/nextcloud/documentation/pull/8209
feel free to approve that one

**Comment by r2evans** [comment]
> 

@fuse314, your repo change to `cron.php` is no longer available. Are you able to post your changes?

**Comment by fuse314** [comment]
@r2evans the equivalent of my changes have been merged with https://github.com/nextcloud/server/pull/32109
The webcron solution currently works for me (I use Nextcloud 23), so I deleted my fork.

**Comment by r2evans** [comment]
The webcron solution is notable _not_ working for me in 24.0.5, not sure if it just fell behind and never caught up or if it is still fundamentally unable to handle the load. (Low user-count in my case.) My workaround is unfortunately to configure my synology to run `docker exec nextcloud php -f cron.php` instead of `curl https://myhost.com/nextcloud/cron.php`: it requires root in cron (discouraged) but has the benefit of doing a "full" cron run instead of the webcron method (whether still hobbled or not). Thanks @fuse314
